In principle, trees should be constructed such that you could null all the types after any phase and retype the trees and get the same types back.  At least, that's what I hear.  Presumably Typed(tree, tpt) might actually live up to that ambition, and "tree setType tpe" would not.  Supposedly this is what -Ycheck does; you will not be surprised to find that -Ycheck:all has not often been found in the "working flawlessly" circle.  Once long ago I got it there, but without it being part of the build, forget it.

It's always interesting to see how it fails though.  Random example:

```
% scalac3 -J-Xmx1g -Ycheck:all -d /tmp ./src/compiler/scala/tools/nsc/typechecker/*.scala
[Not checkable: parser]
[Not checkable: namer]
[Not checkable: packageobjects]
[Now checking: typer]

./src/compiler/scala/tools/nsc/typechecker/Checkable.scala:289: error: redefinition of hashCode method. See SIP-15, criterion 4. is not allowed in value class
private[typechecker] final class Checkability(val value: Int) extends AnyVal { }
                                 ^
./src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:1383: error: 
**** ERROR DURING INTERNAL CHECKING ****
type mismatch;
 found   : Class[(some other)?0(in method traverse)] where type (some other)?0(in method traverse) <: PatternMatching.this.global.Tree
 required: Class[?0(in method traverse)] where type ?0(in method traverse) <: PatternMatching.this.global.Tree
          patmatDebug("untouched "+ (t, t.getClass, t.symbol.ownerChain, currentOwner.ownerChain))
                                          ^
./src/compiler/scala/tools/nsc/typechecker/Typers.scala:2792: error: 
**** ERROR DURING INTERNAL CHECKING ****
type mismatch;
 found   : Class[(some other)?0(in method checkNoDoubleDefs)] where type (some other)?0(in method checkNoDoubleDefs) <: Typers.this.global.Symbol
 required: Class[?0(in method checkNoDoubleDefs)] where type ?0(in method checkNoDoubleDefs) <: Typers.this.global.Symbol
                    ((e1.sym.getClass, e1.sym.info, e1.sym.ownerChain)))
                             ^
three errors found
```

--- a/src/compiler/scala/tools/nsc/typechecker/Typers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
@@ -2602,7 +2602,9 @@ trait Typers extends Modes with Adaptations with Tags {
           }
           // Don't leak implementation details into the type, see SI-6575
           if (isPartial && !typedBlock.isErrorTyped)
-            typedBlock modifyType (_ baseType PartialFunctionClass)
+            typedPos(tree.pos, mode, pt) {
+              Typed(typedBlock, TypeTree(typedBlock.tpe baseType PartialFunctionClass))
+            }
           else typedBlock
         }
     }

